File: app/src/main/java/com/example/dlna/BinaryLightServer.java
package com.example.dlna;

import org.fourthline.cling.UpnpService;
import org.fourthline.cling.UpnpServiceImpl;
import org.fourthline.cling.binding.*;
import org.fourthline.cling.binding.annotations.*;
import org.fourthline.cling.model.*;
import org.fourthline.cling.model.meta.*;
import org.fourthline.cling.model.types.*;

import java.net.HttpURLConnection;
import java.net.URLStreamHandler;
import java.net.URLStreamHandlerFactory;
import java.io.IOException;
import java.net.URL;

public class BinaryLightServer implements Runnable {


    public static class CustomStreamHandlerFactory implements URLStreamHandlerFactory {
        @Override
        public URLStreamHandler createURLStreamHandler(String protocol) {
            // 只处理 HTTP 协议
            if ("http".equals(protocol)) {
                return new java.net.URLStreamHandler() {
                    @Override
                    protected java.net.URLConnection openConnection(URL u) {
                        return new HttpURLConnection(u) {
                            @Override
                            public void disconnect() {
                                System.err.println("BinaryLightServer main start: disconnect");

                            }

                            @Override
                            public boolean usingProxy() {
                                System.err.println("BinaryLightServer main start: usingProxy");

                                return false;
                            }

                            @Override
                            public void connect() {
                                // 这里可以添加自定义连接逻辑
                                System.err.println("BinaryLightServer main start: connect");

                            }
                        };
                    }
                };
            }
            return null; // 返回 null，表示其他协议使用默认处理器
        }
    }

    public static void main(String[] args) throws Exception {
        URL.setURLStreamHandlerFactory(new CustomStreamHandlerFactory());
        System.err.println("BinaryLightServer main start: ");

        // Start a user thread that runs the UPnP stack
        Thread serverThread = new Thread(new BinaryLightServer());
        serverThread.setDaemon(false);
        serverThread.start();
    }

    public void run() {
        try {

            final UpnpService upnpService = new UpnpServiceImpl();

            Runtime.getRuntime().addShutdownHook(new Thread() {
                @Override
                public void run() {
                    upnpService.shutdown();
                }
            });

            // Add the bound local device to the registry
            upnpService.getRegistry().addDevice(createDevice());

        } catch (Exception ex) {
            System.err.println("Exception occured: " + ex);
            ex.printStackTrace(System.err);
            System.exit(1);
        }
    }

    LocalDevice createDevice() throws ValidationException, LocalServiceBindingException, IOException {

        DeviceIdentity identity = new DeviceIdentity(UDN.uniqueSystemIdentifier("Demo Binary Light"));

        DeviceType type = new UDADeviceType("BinaryLight", 1);

        DeviceDetails details = new DeviceDetails("Friendly Binary Light", new ManufacturerDetails("ACME"), new ModelDetails("BinLight2000", "A demo light with on/off switch.", "v1"));

//        Icon icon = new Icon("image/png", 48, 48, 8, getClass().getResource("icon.png"));

        LocalService<SwitchPower> switchPowerService = new AnnotationLocalServiceBinder().read(SwitchPower.class);

        switchPowerService.setManager(new DefaultServiceManager(switchPowerService, SwitchPower.class));

        return new LocalDevice(identity, type, details, switchPowerService);

    /* Several services can be bound to the same device:
    return new LocalDevice(
            identity, type, details, icon,
            new LocalService[] {switchPowerService, myOtherService}
    );
    */

    }

}

File: app/src/main/java/com/example/dlna/BinaryLightServer.kt
package com.example.dlna

import android.util.Log
import org.fourthline.cling.UpnpService
import org.fourthline.cling.UpnpServiceImpl
import org.fourthline.cling.binding.LocalServiceBindingException
import org.fourthline.cling.binding.annotations.AnnotationLocalServiceBinder
import org.fourthline.cling.model.DefaultServiceManager
import org.fourthline.cling.model.ValidationException
import org.fourthline.cling.model.meta.DeviceDetails
import org.fourthline.cling.model.meta.DeviceIdentity
import org.fourthline.cling.model.meta.Icon
import org.fourthline.cling.model.meta.LocalDevice
import org.fourthline.cling.model.meta.LocalService
import org.fourthline.cling.model.meta.ManufacturerDetails
import org.fourthline.cling.model.meta.ModelDetails
import org.fourthline.cling.model.types.UDADeviceType
import org.fourthline.cling.model.types.UDN
import java.io.IOException


class BinaryLightServer : Runnable {
    override fun run() {
        try {
            val upnpService: UpnpService = UpnpServiceImpl()

            Runtime.getRuntime().addShutdownHook(object : Thread() {
                override fun run() {
                    Log.i("MCLOG", "ShutdownHook")
                    upnpService.shutdown()
                }
            })

            // Add the bound local device to the registry
            upnpService.registry.addDevice(
                createDevice()
            )
        } catch (ex: Exception) {
            System.err.println("Exception occured: $ex")
            ex.printStackTrace(System.err)
            System.exit(1)
        }
    }

    @Throws(ValidationException::class, LocalServiceBindingException::class, IOException::class)
    fun createDevice(): LocalDevice {
        val identity = DeviceIdentity(
            UDN.uniqueSystemIdentifier("Demo Binary Light")
        )

        val type = UDADeviceType("BinaryLight", 1)

        val details = DeviceDetails(
            "Friendly Binary Light",
            ManufacturerDetails("ACME"),
            ModelDetails(
                "BinLight2000",
                "A demo light with on/off switch.",
                "v1"
            )
        )

        val icon = Icon("image/png", 48, 48, 8, javaClass.getResource("icon.png"))

        @Suppress("UNCHECKED_CAST")
        val switchPowerService =
            AnnotationLocalServiceBinder().read(SwitchPower::class.java) as LocalService<SwitchPower>
        switchPowerService.setManager(
            DefaultServiceManager(
                switchPowerService,
                SwitchPower::class.java
            )
        )

        return LocalDevice(identity, type, details, icon, switchPowerService)

        // 你也可以使用下面的代码将多个服务绑定到同一个设备:
        // return LocalDevice(
        //     identity, type, details, icon,
        //     arrayOf(switchPowerService, myOtherService)
        // )
    }

}